CARDS-1582: Proms mode: Clinicians can delete only forms that they created

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/CARDS-1582)

This PR causes users in the TrustedUsers group in Proms mode to be unable to delete Forms unless they themselves created it.

This also fixes a bug in the DeleteServlet where deleting any versionable node whose parent is inaccessible to the current user would crash with an Access Denied exception.

**To test**:
- Start CARDS under proms mode with trusted permissions active (`./start_cards.sh --dev -P cards4proms --permissions trusted`)
- Create multiple users from the Admin page (Administration->Users->Create)
- Add them all to the TrustedUsers group (Administration->Groups)
- Create Patients/Visits for testing with the Admin account
- Create Visit Information forms while logged in as the new TrustedUsers
- You should only be allowed to delete that form with the user that created it. The other users should receive an error like below.

![image](https://user-images.githubusercontent.com/4656440/152880759-2354e18a-bc68-456f-9558-f57abcf2ed5d.png)

OrdiNeu committed:
CARDS-1582: Proms mode: Clinicians can delete only forms that they
created

OrdiNeu committed:
CARDS-1582: Fix the Delete Servlet cause a Null Reference Exception when
the parent of a node is inaccessible

OrdiNeu committed:
CARDS-1582: Fix the dialogue displayed after being unable to delete a
form

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -78,6 +75,10 @@ function DeleteButton(props) {
       setEntryNotFound(true);
       openError();
+    } else if (status === 401 || response.url.startsWith(window.location.origin + "/login")) {
+      setErrorMessage(`The currently signed in user does not have permission to delete ${entryName}.`);
```
The message is technically correct, but a bit too technical. I would suggest something simpler instead:
```suggestion
      setErrorMessage(`You do not have permission to delete ${entryName}.`);
```

sashaandjic commented:
The login issues that was resolved in 1475 keeps poping up and making me unable to test this PR properly.
Can we include the changes done in 1475 here too?

marta- commented:
> The login issues that was resolved in 1475 keeps popping up and making me unable to test this PR properly. Interestingly, initially I was able to login as trusted user, but I cold not see any forms created by admin other than Visit info forms. Can we include the changes done in 1475 here too?

@sashaandjic Rebased, please try now.

sashaandjic commented:
I'm still seeing the old wording in error message:
![image](https://user-images.githubusercontent.com/23243244/152930288-4fbb862b-baaa-410e-8097-e35a9a0cbb07.png)
![image](https://user-images.githubusercontent.com/23243244/152930356-818e2e51-4f10-4bd5-afeb-d407c8dd677d.png)

Everything else is fine.

marta- commented:
> I'm still seeing the old wording in error message instead of suggested "`You do not have permission to delete..."

That's expected, i didn't actually change anything related to the 1582 implementation, only rebased for smoother testing.

OrdiNeu committed:
CARDS-1582: Split the error code back into 401/403 depending on user
logged-in status, and handle in DeleteButton instead of changing
fetchWithReLogin

OrdiNeu committed:
CARDS-1582: Add UI change as suggested by @marta-

marta- committed:
CARDS-1582: Proms mode: Clinicians can delete only forms that they created
Revert artifact changes

marta- commented on modules/permissions/src/main/java/io/uhndata/cards/permissions/internal/NotCreatedByRestrictionPattern.java:
```diff
@@ -0,0 +1,79 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.uhndata.cards.permissions.internal;
```
@sdumitriu should this be in `proms-resources` instead?

marta- commented on modules/permissions/src/main/java/io/uhndata/cards/permissions/internal/NotCreatedByRestrictionFactory.java:
```diff
@@ -0,0 +1,78 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.uhndata.cards.permissions.internal;
```
@sdumitriu should this be in `proms-resources` instead?

sdumitriu replied:
No.

sashaandjic commented:
Eventually it did start:
![image](https://user-images.githubusercontent.com/23243244/153025483-b0c19416-6490-42b4-be5c-12c1ab19fd68.png)

sdumitriu reviewed (changes requested):
Currently:
- allow everything
- deny removeNode if notCreatedBy

I would like instead:
- allow everything
- deny removeNode
- allow removeNode if createdBy

OrdiNeu committed:
CARDS-1582: Replace the NotCreatedByRestriction with a
CreatedByRestriction

OrdiNeu commented:
@sdumitriu 
> I would like instead:
Fixed in `8803683` :heavy_check_mark:

sashaandjic commented:
not sure if this is out of scope, but trusted users cannot view survey forms created by someone else
![trusted user](https://user-images.githubusercontent.com/23243244/153077894-bef0e04e-24f0-4dc9-ac73-41a14dfbe0b4.png)
![admin](https://user-images.githubusercontent.com/23243244/153077898-d9fcde2c-9d5d-41ca-814d-7243be51891c.png)

marta- commented:
> not sure if this is out of scope, but trusted users cannot view survey forms created by someone else

Rebased on latest dev just in case.

sashaandjic reviewed (approved):
Tested!

sdumitriu committed:
CARDS-1582: Proms mode: Clinicians can delete only forms that they created
Bugfix: editing forms that someone else created is not allowed

sdumitriu committed:
[misc] Fixed copy-paste wrong comment

sashaandjic reviewed (approved):
tested
